// @vkontakte/api-schema-typescript/dist/index.js
/**
 * This is auto-generated file, don't modify this file manually
 */
export { API_VERSION, } from './common/common';
export { AdsRemoveTargetContactsResponseResultEnumNames, } from './methods/ads';
export { AuthRestoreResponseSuccessEnumNames, } from './methods/auth';
export { FriendsAddResponseEnumNames, FriendsDeleteResponseFriendDeletedEnumNames, FriendsDeleteResponseInRequestDeletedEnumNames, FriendsDeleteResponseOutRequestDeletedEnumNames, FriendsDeleteResponseSuggestionDeletedEnumNames, } from './methods/friends';
export { AdsAdApprovedEnumNames, } from './objects/ads/AdsAdApproved';
export { AdsAdCostTypeEnumNames, } from './objects/ads/AdsAdCostType';
export { AdsAdStatusEnumNames, } from './objects/ads/AdsAdStatus';
export { AdsCampaignStatusEnumNames, } from './objects/ads/AdsCampaignStatus';
export { AdsCriteriaSexEnumNames, } from './objects/ads/AdsCriteriaSex';
export { AppsAppLeaderboardTypeEnumNames, } from './objects/apps/AppsAppLeaderboardType';
export { BaseSexEnumNames, } from './objects/base/BaseSex';
export { BoardDefaultOrderEnumNames, } from './objects/board/BoardDefaultOrder';
export { BugtrackerBugreportStatusAutoUpdateReasonEnumNames, } from './objects/bugtracker/BugtrackerBugreport';
export { FriendsFriendStatusStatusEnumNames, } from './objects/friends/FriendsFriendStatusStatus';
export { GiftsGiftPrivacyEnumNames, } from './objects/gifts/GiftsGiftPrivacy';
export { GroupsBanInfoReasonEnumNames, } from './objects/groups/GroupsBanInfoReason';
export { GroupsGroupAccessEnumNames, } from './objects/groups/GroupsGroupAccess';
export { GroupsGroupAdminLevelEnumNames, } from './objects/groups/GroupsGroupAdminLevel';
export { GroupsGroupAgeLimitsEnumNames, } from './objects/groups/GroupsGroupAgeLimits';
export { GroupsGroupAudioEnumNames, } from './objects/groups/GroupsGroupAudio';
export { GroupsGroupDocsEnumNames, } from './objects/groups/GroupsGroupDocs';
export { GroupsGroupFullWallEnumNames, } from './objects/groups/GroupsGroupFull';
export { GroupsGroupFullAgeLimitsEnumNames, } from './objects/groups/GroupsGroupFullAgeLimits';
export { GroupsGroupFullMemberStatusEnumNames, } from './objects/groups/GroupsGroupFullMemberStatus';
export { GroupsGroupFullSectionEnumNames, } from './objects/groups/GroupsGroupFullSection';
export { GroupsGroupIsClosedEnumNames, } from './objects/groups/GroupsGroupIsClosed';
export { GroupsGroupPhotosEnumNames, } from './objects/groups/GroupsGroupPhotos';
export { GroupsGroupSuggestedPrivacyEnumNames, } from './objects/groups/GroupsGroupSuggestedPrivacy';
export { GroupsGroupTopicsEnumNames, } from './objects/groups/GroupsGroupTopics';
export { GroupsGroupVideoEnumNames, } from './objects/groups/GroupsGroupVideo';
export { GroupsGroupWallEnumNames, } from './objects/groups/GroupsGroupWall';
export { GroupsGroupWikiEnumNames, } from './objects/groups/GroupsGroupWiki';
export { MarketMarketAlbumTypeEnumNames, } from './objects/market/MarketMarketAlbum';
export { MarketMarketItemAvailabilityEnumNames, } from './objects/market/MarketMarketItemAvailability';
export { MarketPricePriceTypeEnumNames, MarketPricePriceUnitEnumNames, } from './objects/market/MarketPrice';
export { MarketServicesViewTypeEnumNames, } from './objects/market/MarketServicesViewType';
export { NewsfeedCommentsItemTypePostTopicIdEnumNames, } from './objects/newsfeed/NewsfeedCommentsItemTypePost';
export { NewsfeedItemWallpostTopicIdEnumNames, } from './objects/newsfeed/NewsfeedItemWallpost';
export { NotificationsSendMessageErrorCodeEnumNames, } from './objects/notifications/NotificationsSendMessageError';
export { OwnerStateStateEnumNames, } from './objects/owner/OwnerState';
export { PagesPrivacySettingsEnumNames, } from './objects/pages/PagesPrivacySettings';
export { UsersUserRelationEnumNames, } from './objects/users/UsersUserRelation';
export { WallWallpostFullTopicIdEnumNames, } from './objects/wall/WallWallpostFull';
