// @gramio/contexts/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityType = exports.PollType = exports.ChatType = void 0;
/** Enum of ChatType property */
var ChatType;
(function (ChatType) {
    ChatType["Private"] = "private";
    ChatType["Group"] = "group";
    ChatType["Supergroup"] = "supergroup";
    ChatType["Channel"] = "channel";
})(ChatType || (exports.ChatType = ChatType = {}));
/** Enum of PollType property */
var PollType;
(function (PollType) {
    PollType["Regular"] = "regular";
    PollType["Quiz"] = "quiz";
})(PollType || (exports.PollType = PollType = {}));
/** Enum of EntityType property */
var EntityType;
(function (EntityType) {
    EntityType["Mention"] = "mention";
    EntityType["Hashtag"] = "hashtag";
    EntityType["Cashtag"] = "cashtag";
    EntityType["BotCommand"] = "bot_command";
    EntityType["Url"] = "url";
    EntityType["Email"] = "email";
    EntityType["PhoneNumber"] = "phone_number";
    EntityType["Bold"] = "bold";
    EntityType["Italic"] = "italic";
    EntityType["Underline"] = "underline";
    EntityType["Strikethrough"] = "strikethrough";
    EntityType["Code"] = "code";
    EntityType["Pre"] = "pre";
    EntityType["TextLink"] = "text_link";
    EntityType["TextMention"] = "text_mention";
})(EntityType || (exports.EntityType = EntityType = {}));
