// @gramio/contexts/dist/structures/attachments/audio.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AudioAttachment = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const photo_size_1 = require("../photo-size");
const utils_1 = require("../../utils");
const file_attachment_1 = require("./file-attachment");
/**
 * This object represents an audio file to be treated as music by the Telegram
 * clients.
 */
// TODO: extended: ['fileId', 'fileUniqueId']
let AudioAttachment = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = file_attachment_1.FileAttachment;
    let _instanceExtraInitializers = [];
    let _get_duration_decorators;
    let _get_performer_decorators;
    let _get_title_decorators;
    let _get_fileName_decorators;
    let _get_mimeType_decorators;
    let _get_fileSize_decorators;
    let _get_thumbnail_decorators;
    var AudioAttachment = class extends _classSuper {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _get_duration_decorators = [(0, inspectable_1.Inspect)()];
            _get_performer_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_title_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_fileName_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_mimeType_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_fileSize_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_thumbnail_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            tslib_1.__esDecorate(this, null, _get_duration_decorators, { kind: "getter", name: "duration", static: false, private: false, access: { has: obj => "duration" in obj, get: obj => obj.duration }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_performer_decorators, { kind: "getter", name: "performer", static: false, private: false, access: { has: obj => "performer" in obj, get: obj => obj.performer }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_title_decorators, { kind: "getter", name: "title", static: false, private: false, access: { has: obj => "title" in obj, get: obj => obj.title }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_fileName_decorators, { kind: "getter", name: "fileName", static: false, private: false, access: { has: obj => "fileName" in obj, get: obj => obj.fileName }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_mimeType_decorators, { kind: "getter", name: "mimeType", static: false, private: false, access: { has: obj => "mimeType" in obj, get: obj => obj.mimeType }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_fileSize_decorators, { kind: "getter", name: "fileSize", static: false, private: false, access: { has: obj => "fileSize" in obj, get: obj => obj.fileSize }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_thumbnail_decorators, { kind: "getter", name: "thumbnail", static: false, private: false, access: { has: obj => "thumbnail" in obj, get: obj => obj.thumbnail }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            AudioAttachment = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        attachmentType = (tslib_1.__runInitializers(this, _instanceExtraInitializers), "audio");
        /** Duration of the audio in seconds as defined by sender */
        get duration() {
            return this.payload.duration;
        }
        /** Performer of the audio as defined by sender or by audio tags */
        get performer() {
            return this.payload.performer;
        }
        /** Title of the audio as defined by sender or by audio tags */
        get title() {
            return this.payload.title;
        }
        /** Original filename as defined by sender */
        get fileName() {
            return this.payload.file_name;
        }
        /** MIME type of the file as defined by sender */
        get mimeType() {
            return this.payload.mime_type;
        }
        /** File size */
        get fileSize() {
            return this.payload.file_size;
        }
        /** Thumbnail of the album cover to which the music file belongs */
        get thumbnail() {
            const { thumbnail } = this.payload;
            if (!thumbnail)
                return undefined;
            return new photo_size_1.PhotoSize(thumbnail);
        }
    };
    return AudioAttachment = _classThis;
})();
exports.AudioAttachment = AudioAttachment;
(0, utils_1.memoizeGetters)(AudioAttachment, ["thumbnail"]);
