// @gramio/contexts/dist/structures/attachments/document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentAttachment = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const utils_1 = require("../../utils");
const photo_size_1 = require("../photo-size");
const file_attachment_1 = require("./file-attachment");
/**
 * This object represents a general file (as opposed to photos, voice messages
 * and audio files).
 */
// TODO: extended: ['fileId', 'fileUniqueId']
let DocumentAttachment = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = file_attachment_1.FileAttachment;
    let _instanceExtraInitializers = [];
    let _get_thumbnail_decorators;
    let _get_fileName_decorators;
    let _get_mimeType_decorators;
    let _get_fileSize_decorators;
    var DocumentAttachment = class extends _classSuper {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _get_thumbnail_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_fileName_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_mimeType_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_fileSize_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            tslib_1.__esDecorate(this, null, _get_thumbnail_decorators, { kind: "getter", name: "thumbnail", static: false, private: false, access: { has: obj => "thumbnail" in obj, get: obj => obj.thumbnail }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_fileName_decorators, { kind: "getter", name: "fileName", static: false, private: false, access: { has: obj => "fileName" in obj, get: obj => obj.fileName }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_mimeType_decorators, { kind: "getter", name: "mimeType", static: false, private: false, access: { has: obj => "mimeType" in obj, get: obj => obj.mimeType }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_fileSize_decorators, { kind: "getter", name: "fileSize", static: false, private: false, access: { has: obj => "fileSize" in obj, get: obj => obj.fileSize }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            DocumentAttachment = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        attachmentType = (tslib_1.__runInitializers(this, _instanceExtraInitializers), "document");
        /** Document thumbnail as defined by sender */
        get thumbnail() {
            const { thumbnail } = this.payload;
            if (!thumbnail) {
                return undefined;
            }
            return new photo_size_1.PhotoSize(thumbnail);
        }
        /** Original filename as defined by sender */
        get fileName() {
            return this.payload.file_name;
        }
        /** MIME type of the file as defined by sender */
        get mimeType() {
            return this.payload.mime_type;
        }
        /** File size */
        get fileSize() {
            return this.payload.file_size;
        }
    };
    return DocumentAttachment = _classThis;
})();
exports.DocumentAttachment = DocumentAttachment;
(0, utils_1.memoizeGetters)(DocumentAttachment, ["thumbnail"]);
