// @gramio/contexts/dist/structures/attachments/photo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhotoAttachment = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const attachment_1 = require("./attachment");
/** This object represents a photo file with it's sizes */
let PhotoAttachment = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = attachment_1.Attachment;
    let _instanceExtraInitializers = [];
    let _get_bigSize_decorators;
    let _get_mediumSize_decorators;
    let _get_smallSize_decorators;
    var PhotoAttachment = class extends _classSuper {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _get_bigSize_decorators = [(0, inspectable_1.Inspect)()];
            _get_mediumSize_decorators = [(0, inspectable_1.Inspect)()];
            _get_smallSize_decorators = [(0, inspectable_1.Inspect)()];
            tslib_1.__esDecorate(this, null, _get_bigSize_decorators, { kind: "getter", name: "bigSize", static: false, private: false, access: { has: obj => "bigSize" in obj, get: obj => obj.bigSize }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_mediumSize_decorators, { kind: "getter", name: "mediumSize", static: false, private: false, access: { has: obj => "mediumSize" in obj, get: obj => obj.mediumSize }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_smallSize_decorators, { kind: "getter", name: "smallSize", static: false, private: false, access: { has: obj => "smallSize" in obj, get: obj => obj.smallSize }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            PhotoAttachment = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        sorted;
        attachmentType = "photo";
        constructor(payload) {
            super();
            this.payload = payload;
            this.sorted = [...this.payload].sort((first, second) => second.width * second.height - first.width * first.height);
        }
        /** Photo sizes */
        get sizes() {
            return this.payload;
        }
        /** Biggest size of the photo */
        get bigSize() {
            return this.sorted[0];
        }
        /** Medium size of the photo */
        get mediumSize() {
            return this.sorted[Math.floor(this.sorted.length / 2)];
        }
        /** Smallest size of the photo */
        get smallSize() {
            return this.sorted[this.sorted.length - 1];
        }
    };
    return PhotoAttachment = _classThis;
})();
exports.PhotoAttachment = PhotoAttachment;
