// @gramio/contexts/dist/contexts/mixins/chat-invite-control.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatInviteControlMixin = void 0;
/** This object represents a mixin that works with all `*ChatInviteLink` methods */
class ChatInviteControlMixin {
    /** Generates new primary invite link */
    exportInviteLink(params) {
        return this.bot.api.exportChatInviteLink({
            chat_id: this.chatId,
            ...params,
        });
    }
    /** Creates an additional invite link */
    createInviteLink(params) {
        return this.bot.api.createChatInviteLink({
            chat_id: this.chatId,
            ...params,
        });
    }
    /** Edits non-primary invite link created by the bot */
    editInviteLink(link, params) {
        return this.bot.api.editChatInviteLink({
            chat_id: this.chatId,
            invite_link: link,
            ...params,
        });
    }
    /** Revokes an invite link generated by a bot */
    revokeInviteLink(link, params) {
        return this.bot.api.revokeChatInviteLink({
            chat_id: this.chatId,
            invite_link: link,
            ...params,
        });
    }
}
exports.ChatInviteControlMixin = ChatInviteControlMixin;
