// @gramio/contexts/dist/contexts/mixins/chat-control.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatControlMixin = void 0;
/** This object represents a mixin that is responsible for all the chat management methods */
class ChatControlMixin {
    /** Sets a custom title */
    setCustomTitle(title, params) {
        return this.bot.api.setChatAdministratorCustomTitle({
            chat_id: this.chatId,
            user_id: this.senderId,
            custom_title: title,
            ...params,
        });
    }
    /** Sets default chat permissions */
    setChatDefaultPermissions(permissions, params) {
        return this.bot.api.setChatPermissions({
            chat_id: this.chatId,
            permissions,
            ...params,
        });
    }
    /** Sets a new profile photo for the chat */
    setChatPhoto(photo, params) {
        return this.bot.api.setChatPhoto({
            chat_id: this.chatId,
            photo,
            ...params,
        });
    }
    /** Deletes a chat photo */
    // INFO: had to rename it from `deleteChatPhoto` because of `TelegramMessage.delete_chat_photo` 😤😤
    removeChatPhoto(params) {
        return this.bot.api.deleteChatPhoto({
            chat_id: this.chatId,
            ...params,
        });
    }
    /** Changes chat title */
    setChatTitle(title, params) {
        return this.bot.api.setChatTitle({
            chat_id: this.chatId,
            title,
            ...params,
        });
    }
    /** Changes chat description */
    setChatDescription(description, params) {
        return this.bot.api.setChatDescription({
            chat_id: this.chatId,
            description,
            ...params,
        });
    }
    /** Sets new group stickerset */
    setChatStickerSet(name, params) {
        return this.bot.api.setChatStickerSet({
            chat_id: this.chatId,
            sticker_set_name: name,
            ...params,
        });
    }
    /** Deletes group stickerset */
    deleteChatStickerSet(params) {
        return this.bot.api.deleteChatStickerSet({
            chat_id: this.chatId,
            ...params,
        });
    }
}
exports.ChatControlMixin = ChatControlMixin;
