// @gramio/contexts/dist/contexts/boost-added.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BoostAddedContext = void 0;
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const inspectable_1 = require("inspectable");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/** This object represents a service message about a forum topic closed in the chat. Currently holds no information. */
class BoostAddedContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    /** Create new BoostAddedContext */
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "boost_added",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** Number of boosts added by the user */
    get boostCount() {
        return this.payload.boost_added?.boost_count || 0;
    }
}
exports.BoostAddedContext = BoostAddedContext;
(0, utils_1.applyMixins)(BoostAddedContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.ChatInviteControlMixin,
    index_2.ChatControlMixin,
    index_2.ChatSenderControlMixin,
    index_2.ChatMemberControlMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, inspectable_1.inspectable)(BoostAddedContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            boostCount: context.boostCount,
        };
    },
});
