// @gramio/contexts/dist/contexts/business-connection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BusinessConnectionContext = void 0;
const inspectable_1 = require("inspectable");
const business_connection_1 = require("../structures/business-connection");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_1 = require("./mixins/index");
/** This object  Describes the connection of the bot with a business account. */
class BusinessConnectionContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    /** Create new BusinessConnectionContext */
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "business_connection",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
}
exports.BusinessConnectionContext = BusinessConnectionContext;
(0, utils_1.applyMixins)(BusinessConnectionContext, [business_connection_1.BusinessConnection, index_1.CloneMixin]);
(0, inspectable_1.inspectable)(BusinessConnectionContext, {
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    serialize(context) {
        return context;
    },
});
