// @gramio/contexts/dist/contexts/business-messages-deleted.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BusinessMessagesDeletedContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/** This object represents a boost added to a chat or changed. */
class BusinessMessagesDeletedContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    /** Create new BusinessMessagesDeletedContext */
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "deleted_business_messages",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
}
exports.BusinessMessagesDeletedContext = BusinessMessagesDeletedContext;
(0, utils_1.applyMixins)(BusinessMessagesDeletedContext, [
    index_1.BusinessMessagesDeleted,
    index_2.CloneMixin,
]);
(0, inspectable_1.inspectable)(BusinessMessagesDeletedContext, {
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    serialize(context) {
        return context;
    },
});
