// @gramio/contexts/dist/contexts/chat-background-set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatBackgroundSetContext = void 0;
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const inspectable_1 = require("inspectable");
const background_type_1 = require("../structures/background-type");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/** This object represents a service message about chat background set. */
class ChatBackgroundSetContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "chat_background_set",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** Type of the background */
    get type() {
        return new background_type_1.backgroundTypeMap[this.payload.chat_background_set.type.type](
        // @ts-expect-error
        this.payload.chat_background_set.type);
    }
}
exports.ChatBackgroundSetContext = ChatBackgroundSetContext;
(0, utils_1.applyMixins)(ChatBackgroundSetContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.ChatInviteControlMixin,
    index_2.ChatControlMixin,
    index_2.ChatSenderControlMixin,
    index_2.ChatMemberControlMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, inspectable_1.inspectable)(ChatBackgroundSetContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            type: context.type,
        };
    },
});
