// @gramio/contexts/dist/contexts/chat-boost.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatBoostContext = void 0;
const chat_boost_updated_1 = require("../structures/chat-boost-updated");
const utils_1 = require("../utils");
const inspectable_1 = require("inspectable");
const context_1 = require("./context");
const index_1 = require("./mixins/index");
/** This object represents a boost added to a chat or changed. */
class ChatBoostContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "chat_boost",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
}
exports.ChatBoostContext = ChatBoostContext;
(0, utils_1.applyMixins)(ChatBoostContext, [chat_boost_updated_1.ChatBoostUpdated, index_1.SendMixin, index_1.CloneMixin]);
(0, inspectable_1.inspectable)(ChatBoostContext, {
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    serialize(context) {
        const payload = {
            chat: context.chat,
            boost: context.boost,
        };
        return payload;
    },
});
