// @gramio/contexts/dist/contexts/forum-topic-created.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForumTopicCreatedContext = void 0;
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const inspectable_1 = require("inspectable");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/** This object represents a service message about a new forum topic created in the chat. */
class ForumTopicCreatedContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    event;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "forum_topic_created",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
        this.event = this.payload
            .forum_topic_created;
    }
    /** Name of the topic */
    get name() {
        return this.event.name;
    }
    /** Color of the topic icon in RGB format */
    get iconColor() {
        return this.event.icon_color;
    }
    /** Unique identifier of the custom emoji shown as the topic icon */
    get iconCustomEmojiId() {
        return this.event.icon_custom_emoji_id;
    }
    /** Checks whether the event has `iconCustomEmojiId` property */
    hasIconCustomEmojiId() {
        return this.iconCustomEmojiId !== undefined;
    }
}
exports.ForumTopicCreatedContext = ForumTopicCreatedContext;
(0, utils_1.applyMixins)(ForumTopicCreatedContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.ForumMixin,
    index_2.ChatInviteControlMixin,
    index_2.ChatControlMixin,
    index_2.ChatSenderControlMixin,
    index_2.ChatMemberControlMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, inspectable_1.inspectable)(ForumTopicCreatedContext, {
    serialize(context) {
        const payload = {
            name: context.name,
            iconColor: context.iconColor,
            iconCustomEmojiId: context.iconCustomEmojiId,
        };
        return (0, utils_1.filterPayload)(payload);
    },
});
