// @gramio/contexts/dist/contexts/general-forum-topic-unhidden.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeneralForumTopicUnhiddenContext = void 0;
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const inspectable_1 = require("inspectable");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/** This object represents a service message about General forum topic unhidden in the chat. Currently holds no information. */
class GeneralForumTopicUnhiddenContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "general_forum_topic_hidden",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** Checks whether this topic is actually a 'General' one */
    isGeneralTopic() {
        return this.threadId === undefined;
    }
}
exports.GeneralForumTopicUnhiddenContext = GeneralForumTopicUnhiddenContext;
(0, utils_1.applyMixins)(GeneralForumTopicUnhiddenContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.ForumMixin,
    index_2.ChatInviteControlMixin,
    index_2.ChatControlMixin,
    index_2.ChatSenderControlMixin,
    index_2.ChatMemberControlMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, inspectable_1.inspectable)(GeneralForumTopicUnhiddenContext, {
    serialize(context) {
        return {};
    },
});
