// @gramio/contexts/dist/contexts/giveaway-winners.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GiveawayWinnersContext = void 0;
const inspectable_1 = require("inspectable");
const message_1 = require("../structures/message");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_1 = require("./mixins/index");
/** This object represents a message about the completion of a giveaway with public winners. */
class GiveawayWinnersContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "giveaway_winners",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** Giveaway winners */
    get eventGiveaway() {
        return this.giveawayWinners;
    }
}
exports.GiveawayWinnersContext = GiveawayWinnersContext;
(0, utils_1.applyMixins)(GiveawayWinnersContext, [
    message_1.Message,
    index_1.TargetMixin,
    index_1.SendMixin,
    index_1.ChatActionMixin,
    index_1.NodeMixin,
    index_1.ForumMixin,
    index_1.ChatInviteControlMixin,
    index_1.ChatControlMixin,
    index_1.ChatSenderControlMixin,
    index_1.ChatMemberControlMixin,
    index_1.PinsMixin,
    index_1.CloneMixin,
]);
(0, inspectable_1.inspectable)(GiveawayWinnersContext, {
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    serialize(context) {
        const payload = {
            id: context.id,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventGiveaway: context.eventGiveaway,
        };
        return (0, utils_1.filterPayload)(payload);
    },
});
