// @gramio/contexts/dist/contexts/giveaway-created.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GiveawayCreatedContext = void 0;
const inspectable_1 = require("inspectable");
const message_1 = require("../structures/message");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_1 = require("./mixins/index");
/** This object represents a service message about the creation of a scheduled giveaway. Currently holds no information. */
class GiveawayCreatedContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "giveaway_created",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
}
exports.GiveawayCreatedContext = GiveawayCreatedContext;
(0, utils_1.applyMixins)(GiveawayCreatedContext, [
    message_1.Message,
    index_1.TargetMixin,
    index_1.SendMixin,
    index_1.ChatActionMixin,
    index_1.NodeMixin,
    index_1.ForumMixin,
    index_1.ChatInviteControlMixin,
    index_1.ChatControlMixin,
    index_1.ChatSenderControlMixin,
    index_1.ChatMemberControlMixin,
    index_1.PinsMixin,
    index_1.CloneMixin,
]);
(0, inspectable_1.inspectable)(GiveawayCreatedContext, {
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    serialize(context) {
        return {
            id: context.id,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
        };
    },
});
