// @gramio/contexts/dist/contexts/invoice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvoiceContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/**
 * Message is an invoice for a [payment](https://core.telegram.org/bots/api/#payments), information about the invoice. [More about payments »](https://core.telegram.org/bots/api/#payments)
 *
 * [Documentation](https://core.telegram.org/bots/api/#invoice)
 */
class InvoiceContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "invoice",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** Invoice */
    get eventInvoice() {
        return new index_1.Invoice(this.payload.invoice);
    }
}
exports.InvoiceContext = InvoiceContext;
(0, utils_1.applyMixins)(InvoiceContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, utils_1.memoizeGetters)(InvoiceContext, ["eventInvoice"]);
(0, inspectable_1.inspectable)(InvoiceContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventInvoice: context.eventInvoice,
        };
    },
});
