// @gramio/contexts/dist/contexts/message-auto-delete-timer-changed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageAutoDeleteTimerChangedContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/**
 * This object represents a service message about a change in auto-delete timer settings.
 *
 * [Documentation](https://core.telegram.org/bots/api/#messageautodeletetimerchanged)
 */
class MessageAutoDeleteTimerChangedContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "message_auto_delete_timer_changed",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** Message auto delete timer */
    get autoDeleteTimer() {
        return new index_1.MessageAutoDeleteTimerChanged(this.payload
            .message_auto_delete_timer_changed);
    }
}
exports.MessageAutoDeleteTimerChangedContext = MessageAutoDeleteTimerChangedContext;
(0, utils_1.applyMixins)(MessageAutoDeleteTimerChangedContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, utils_1.memoizeGetters)(MessageAutoDeleteTimerChangedContext, ["autoDeleteTimer"]);
(0, inspectable_1.inspectable)(MessageAutoDeleteTimerChangedContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            autoDeleteTimer: context.autoDeleteTimer,
        };
    },
});
