// @gramio/contexts/dist/contexts/new-chat-photo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewChatPhotoContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/** A chat photo was change to this value */
class NewChatPhotoContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "new_chat_photo",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** New chat photo */
    get eventPhoto() {
        const sizes = this.payload
            .new_chat_photo;
        return sizes.map((size) => new index_1.PhotoSize(size));
    }
}
exports.NewChatPhotoContext = NewChatPhotoContext;
(0, utils_1.applyMixins)(NewChatPhotoContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.ChatInviteControlMixin,
    index_2.ChatControlMixin,
    index_2.ChatSenderControlMixin,
    index_2.ChatMemberControlMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, inspectable_1.inspectable)(NewChatPhotoContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventPhoto: context.eventPhoto,
        };
    },
});
