// @gramio/contexts/dist/contexts/new-chat-members.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewChatMembersContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/**  New members that were added to the group or supergroup and information about them (the bot itself may be one of these members) */
class NewChatMembersContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "new_chat_members",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** New chat members */
    get eventMembers() {
        const members = this.payload
            .new_chat_members;
        return members.map((member) => new index_1.User(member));
    }
}
exports.NewChatMembersContext = NewChatMembersContext;
(0, utils_1.applyMixins)(NewChatMembersContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.ChatInviteControlMixin,
    index_2.ChatControlMixin,
    index_2.ChatSenderControlMixin,
    index_2.ChatMemberControlMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, inspectable_1.inspectable)(NewChatMembersContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventMembers: context.eventMembers,
        };
    },
});
