// @gramio/contexts/dist/contexts/new-chat-title.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewChatTitleContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/** A chat title was changed to this value */
class NewChatTitleContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "new_chat_title",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** New chat title */
    get eventTitle() {
        return this.payload.new_chat_title;
    }
}
exports.NewChatTitleContext = NewChatTitleContext;
(0, utils_1.applyMixins)(NewChatTitleContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.ChatInviteControlMixin,
    index_2.ChatControlMixin,
    index_2.ChatSenderControlMixin,
    index_2.ChatMemberControlMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, inspectable_1.inspectable)(NewChatTitleContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventTitle: context.eventTitle,
        };
    },
});
