// @gramio/contexts/dist/contexts/passport-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PassportDataContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/**
 * Describes Telegram Passport data shared with the bot by the user.
 *
 * [Documentation](https://core.telegram.org/bots/api/#passportdata)
 */
class PassportDataContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "passport_data",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** Telegram Passport data */
    get passportData() {
        return new index_1.PassportData(this.payload.passport_data);
    }
}
exports.PassportDataContext = PassportDataContext;
(0, utils_1.applyMixins)(PassportDataContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, utils_1.memoizeGetters)(PassportDataContext, ["passportData"]);
(0, inspectable_1.inspectable)(PassportDataContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            passportData: context.passportData,
        };
    },
});
