// @gramio/contexts/dist/contexts/proximity-alert-triggered.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProximityAlertTriggeredContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/**
 * This object represents the content of a service message, sent whenever a user in the chat triggers a proximity alert set by another user.
 *
 * [Documentation](https://core.telegram.org/bots/api/#proximityalerttriggered)
 */
class ProximityAlertTriggeredContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "proximity_alert_triggered",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /**
     * Service message.
     * A user in the chat triggered another user's proximity alert
     * while sharing Live Location.
     */
    get proximityAlert() {
        return new index_1.ProximityAlertTriggered(this.payload
            .proximity_alert_triggered);
    }
}
exports.ProximityAlertTriggeredContext = ProximityAlertTriggeredContext;
(0, utils_1.applyMixins)(ProximityAlertTriggeredContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.ChatInviteControlMixin,
    index_2.ChatControlMixin,
    index_2.ChatSenderControlMixin,
    index_2.ChatMemberControlMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, utils_1.memoizeGetters)(ProximityAlertTriggeredContext, ["proximityAlert"]);
(0, inspectable_1.inspectable)(ProximityAlertTriggeredContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            proximityAlert: context.proximityAlert,
        };
    },
});
