// @gramio/contexts/dist/contexts/refunded-payment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RefundedPaymentContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const refunded_payment_1 = require("../structures/refunded-payment");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/**
 * This object contains basic information about a successful payment.
 *
 * [Documentation](https://core.telegram.org/bots/api/#RefundedPayment)
 */
class RefundedPaymentContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "refunded_payment",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /**
     * This object contains basic information about a refunded payment.
     *
     * [Documentation](https://core.telegram.org/bots/api/#refundedpayment)
     */
    get eventRefundedPayment() {
        return new refunded_payment_1.RefundedPayment(this.payload.refunded_payment);
    }
}
exports.RefundedPaymentContext = RefundedPaymentContext;
(0, utils_1.applyMixins)(RefundedPaymentContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, utils_1.memoizeGetters)(RefundedPaymentContext, ["eventRefundedPayment"]);
(0, inspectable_1.inspectable)(RefundedPaymentContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventRefundedPayment: context.eventRefundedPayment,
        };
    },
});
