// @gramio/contexts/dist/contexts/removed-chat-boost.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemovedChatBoostContext = void 0;
const inspectable_1 = require("inspectable");
const chat_boost_removed_1 = require("../structures/chat-boost-removed");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_1 = require("./mixins/index");
/** This object represents a boost removed from a chat. */
class RemovedChatBoostContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "removed_chat_boost",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
}
exports.RemovedChatBoostContext = RemovedChatBoostContext;
(0, utils_1.applyMixins)(RemovedChatBoostContext, [chat_boost_removed_1.ChatBoostRemoved, index_1.SendMixin, index_1.CloneMixin]);
(0, inspectable_1.inspectable)(RemovedChatBoostContext, {
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    serialize(context) {
        const payload = {
            id: context.id,
            chat: context.chat,
            removeDate: context.removeDate,
            source: context.source,
        };
        return payload;
    },
});
