// @gramio/contexts/dist/contexts/successful-payment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuccessfulPaymentContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/**
 * This object contains basic information about a successful payment.
 *
 * [Documentation](https://core.telegram.org/bots/api/#successfulpayment)
 */
class SuccessfulPaymentContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "successful_payment",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** Received payment */
    get eventPayment() {
        return new index_1.SuccessfulPayment(this.payload
            .successful_payment);
    }
}
exports.SuccessfulPaymentContext = SuccessfulPaymentContext;
(0, utils_1.applyMixins)(SuccessfulPaymentContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, utils_1.memoizeGetters)(SuccessfulPaymentContext, ["eventPayment"]);
(0, inspectable_1.inspectable)(SuccessfulPaymentContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventPayment: context.eventPayment,
        };
    },
});
