// @gramio/contexts/dist/contexts/users-shared.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsersSharedContext = void 0;
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const inspectable_1 = require("inspectable");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/** This object contains information about the users whose identifiers were shared with the bot using a `KeyboardButtonRequestUsers` button. */
class UsersSharedContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    event;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "users_shared",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
        this.event = this.payload.users_shared;
    }
    /** Identifier of the request */
    get requestId() {
        return this.event.request_id;
    }
    /** Identifier of the shared user. This number may have more than 32 significant bits and some programming languages may have difficulty/silent defects in interpreting it. But it has at most 52 significant bits, so a 64-bit integer or double-precision float type are safe for storing this identifier. The bot may not have access to the user and could be unable to use this identifier, unless the user is already known to the bot by some other means. */
    get users() {
        return this.event.users.map((user) => new index_1.SharedUser(user));
    }
}
exports.UsersSharedContext = UsersSharedContext;
(0, utils_1.applyMixins)(UsersSharedContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, inspectable_1.inspectable)(UsersSharedContext, {
    serialize(context) {
        return {
            requestId: context.requestId,
            users: context.users,
        };
    },
});
