// @gramio/contexts/dist/contexts/video-chat-started.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoChatStartedContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/**
 * This object represents a service message about a video chat started in the chat.
 *
 * [Documentation](https://core.telegram.org/bots/api/#videochatstarted)
 */
class VideoChatStartedContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "video_chat_started",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** Service message: video chat started */
    get eventStarted() {
        return new index_1.VideoChatStarted(this.payload
            .video_chat_started);
    }
}
exports.VideoChatStartedContext = VideoChatStartedContext;
(0, utils_1.applyMixins)(VideoChatStartedContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.ChatInviteControlMixin,
    index_2.ChatControlMixin,
    index_2.ChatSenderControlMixin,
    index_2.ChatMemberControlMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, utils_1.memoizeGetters)(VideoChatStartedContext, ["eventStarted"]);
(0, inspectable_1.inspectable)(VideoChatStartedContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventStarted: context.eventStarted,
        };
    },
});
