// @gramio/contexts/dist/contexts/video-chat-scheduled.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoChatScheduledContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/**
 * This object represents a service message about a video chat scheduled in the chat.
 *
 * [Documentation](https://core.telegram.org/bots/api/#videochatscheduled)
 */
class VideoChatScheduledContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "video_chat_scheduled",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** Service message: video chat scheduled */
    get eventScheduled() {
        return new index_1.VideoChatScheduled(this.payload
            .video_chat_scheduled);
    }
}
exports.VideoChatScheduledContext = VideoChatScheduledContext;
(0, utils_1.applyMixins)(VideoChatScheduledContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.ChatInviteControlMixin,
    index_2.ChatControlMixin,
    index_2.ChatSenderControlMixin,
    index_2.ChatMemberControlMixin,
    index_2.PinsMixin,
    index_2.CloneMixin,
]);
(0, utils_1.memoizeGetters)(VideoChatScheduledContext, ["eventScheduled"]);
(0, inspectable_1.inspectable)(VideoChatScheduledContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventScheduled: context.eventScheduled,
        };
    },
});
