// @gramio/contexts/dist/contexts/web-app-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebAppDataContext = void 0;
const inspectable_1 = require("inspectable");
const index_1 = require("../structures/index");
const utils_1 = require("../utils");
const context_1 = require("./context");
const index_2 = require("./mixins/index");
/**
 * Describes data sent from a [Web App](https://core.telegram.org/bots/webapps) to the bot.
 *
 * [Documentation](https://core.telegram.org/bots/api/#webappdata)
 */
class WebAppDataContext extends context_1.Context {
    /** The raw data that is used for this Context */
    payload;
    constructor(options) {
        super({
            bot: options.bot,
            updateType: "web_app_data",
            updateId: options.updateId,
            update: options.update,
        });
        this.payload = options.payload;
    }
    /** The data. Be aware that a bad client can send arbitrary data in this field. */
    get data() {
        const webAppData = this.payload
            .web_app_data;
        return webAppData.data;
    }
    /**
     * Text of the `web_app` keyboard button, from which the Web App was opened.
     * Be aware that a bad client can send arbitrary data in this field.
     */
    get buttonText() {
        const webAppData = this.payload
            .web_app_data;
        return webAppData.button_text;
    }
}
exports.WebAppDataContext = WebAppDataContext;
(0, utils_1.applyMixins)(WebAppDataContext, [
    index_1.Message,
    index_2.TargetMixin,
    index_2.SendMixin,
    index_2.ChatActionMixin,
    index_2.NodeMixin,
    index_2.CloneMixin,
]);
(0, inspectable_1.inspectable)(WebAppDataContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            data: context.data,
            buttonText: context.buttonText,
        };
    },
});
