// @gramio/contexts/dist/structures/message-origin/channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageOriginChannel = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const chat_1 = require("../chat");
const utils_1 = require("../../utils");
const message_origin_1 = require("./message-origin");
/** The message was originally sent to a channel chat. */
let MessageOriginChannel = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = message_origin_1.MessageOrigin;
    let _instanceExtraInitializers = [];
    let _get_type_decorators;
    let _get_date_decorators;
    let _get_chat_decorators;
    let _get_messageId_decorators;
    let _get_authorSignature_decorators;
    var MessageOriginChannel = class extends _classSuper {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _get_type_decorators = [(0, inspectable_1.Inspect)()];
            _get_date_decorators = [(0, inspectable_1.Inspect)()];
            _get_chat_decorators = [(0, inspectable_1.Inspect)()];
            _get_messageId_decorators = [(0, inspectable_1.Inspect)()];
            _get_authorSignature_decorators = [(0, inspectable_1.Inspect)()];
            tslib_1.__esDecorate(this, null, _get_type_decorators, { kind: "getter", name: "type", static: false, private: false, access: { has: obj => "type" in obj, get: obj => obj.type }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_date_decorators, { kind: "getter", name: "date", static: false, private: false, access: { has: obj => "date" in obj, get: obj => obj.date }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_chat_decorators, { kind: "getter", name: "chat", static: false, private: false, access: { has: obj => "chat" in obj, get: obj => obj.chat }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_messageId_decorators, { kind: "getter", name: "messageId", static: false, private: false, access: { has: obj => "messageId" in obj, get: obj => obj.messageId }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_authorSignature_decorators, { kind: "getter", name: "authorSignature", static: false, private: false, access: { has: obj => "authorSignature" in obj, get: obj => obj.authorSignature }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            MessageOriginChannel = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            super(payload);
            this.payload = payload;
        }
        /** Type of the message origin, always `channel` */
        get type() {
            return this.payload.type;
        }
        /** Date the message was sent originally in Unix time */
        get date() {
            return this.payload.date;
        }
        /** Channel chat to which the message was originally sent */
        get chat() {
            return new chat_1.Chat(this.payload.chat);
        }
        /** Unique message identifier inside the chat */
        get messageId() {
            return this.payload.message_id;
        }
        /** Signature of the original post author */
        get authorSignature() {
            return this.payload.author_signature;
        }
    };
    return MessageOriginChannel = _classThis;
})();
exports.MessageOriginChannel = MessageOriginChannel;
(0, utils_1.memoizeGetters)(MessageOriginChannel, ["chat"]);
