// @gramio/contexts/dist/structures/message-origin/hidden-user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageOriginHiddenUser = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const message_origin_1 = require("./message-origin");
/** The message was originally sent by an unknown user. */
let MessageOriginHiddenUser = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = message_origin_1.MessageOrigin;
    let _instanceExtraInitializers = [];
    let _get_type_decorators;
    let _get_date_decorators;
    let _get_senderUserName_decorators;
    var MessageOriginHiddenUser = class extends _classSuper {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _get_type_decorators = [(0, inspectable_1.Inspect)()];
            _get_date_decorators = [(0, inspectable_1.Inspect)()];
            _get_senderUserName_decorators = [(0, inspectable_1.Inspect)()];
            tslib_1.__esDecorate(this, null, _get_type_decorators, { kind: "getter", name: "type", static: false, private: false, access: { has: obj => "type" in obj, get: obj => obj.type }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_date_decorators, { kind: "getter", name: "date", static: false, private: false, access: { has: obj => "date" in obj, get: obj => obj.date }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_senderUserName_decorators, { kind: "getter", name: "senderUserName", static: false, private: false, access: { has: obj => "senderUserName" in obj, get: obj => obj.senderUserName }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            MessageOriginHiddenUser = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            super(payload);
            this.payload = payload;
        }
        /** Type of the message origin, always `hidden_user` */
        get type() {
            return this.payload.type;
        }
        /** Date the message was sent originally in Unix time */
        get date() {
            return this.payload.date;
        }
        /** Name of the user that sent the message originally */
        get senderUserName() {
            return this.payload.sender_user_name;
        }
    };
    return MessageOriginHiddenUser = _classThis;
})();
exports.MessageOriginHiddenUser = MessageOriginHiddenUser;
