// @gramio/contexts/dist/structures/background-type-wallpaper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackgroundTypeWallpaper = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const utils_1 = require("../utils");
const document_1 = require("./attachments/document");
/**
 * The background is a wallpaper in the JPEG format.
 *
 * [Documentation](https://core.telegram.org/bots/api/#backgroundtypewallpaper)
 */
let BackgroundTypeWallpaper = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_type_decorators;
    let _get_document_decorators;
    let _get_darkThemeDimming_decorators;
    let _get_isBlurred_decorators;
    let _get_isMoving_decorators;
    var BackgroundTypeWallpaper = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_type_decorators = [(0, inspectable_1.Inspect)()];
            _get_document_decorators = [(0, inspectable_1.Inspect)()];
            _get_darkThemeDimming_decorators = [(0, inspectable_1.Inspect)()];
            _get_isBlurred_decorators = [(0, inspectable_1.Inspect)()];
            _get_isMoving_decorators = [(0, inspectable_1.Inspect)()];
            tslib_1.__esDecorate(this, null, _get_type_decorators, { kind: "getter", name: "type", static: false, private: false, access: { has: obj => "type" in obj, get: obj => obj.type }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_document_decorators, { kind: "getter", name: "document", static: false, private: false, access: { has: obj => "document" in obj, get: obj => obj.document }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_darkThemeDimming_decorators, { kind: "getter", name: "darkThemeDimming", static: false, private: false, access: { has: obj => "darkThemeDimming" in obj, get: obj => obj.darkThemeDimming }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_isBlurred_decorators, { kind: "getter", name: "isBlurred", static: false, private: false, access: { has: obj => "isBlurred" in obj, get: obj => obj.isBlurred }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_isMoving_decorators, { kind: "getter", name: "isMoving", static: false, private: false, access: { has: obj => "isMoving" in obj, get: obj => obj.isMoving }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            BackgroundTypeWallpaper = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        /** [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) */
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /**
         * Type of the background, always “wallpaper”
         */
        get type() {
            return this.payload.type;
        }
        /**
         * Document with the wallpaper
         */
        get document() {
            return new document_1.DocumentAttachment(this.payload.document);
        }
        /**
         * Dimming of the background in dark themes, as a percentage; 0-100
         */
        get darkThemeDimming() {
            return this.payload.dark_theme_dimming;
        }
        /**
         * *Optional*. *True*, if the wallpaper is downscaled to fit in a 450x450 square and then box-blurred with radius 12
         */
        get isBlurred() {
            return this.payload.is_blurred;
        }
        /**
         * *Optional*. *True*, if the background moves slightly when the device is tilted
         */
        get isMoving() {
            return this.payload.is_moving;
        }
    };
    return BackgroundTypeWallpaper = _classThis;
})();
exports.BackgroundTypeWallpaper = BackgroundTypeWallpaper;
(0, utils_1.memoizeGetters)(BackgroundTypeWallpaper, ["document"]);
