// @gramio/contexts/dist/structures/background-fill-gradient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackgroundFillGradient = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
/**
 * The background is a gradient fill.
 *
 * [Documentation](https://core.telegram.org/bots/api/#backgroundfillgradient)
 */
let BackgroundFillGradient = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_type_decorators;
    let _get_topColor_decorators;
    let _get_bottomColor_decorators;
    let _get_rotationAngle_decorators;
    var BackgroundFillGradient = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_type_decorators = [(0, inspectable_1.Inspect)()];
            _get_topColor_decorators = [(0, inspectable_1.Inspect)()];
            _get_bottomColor_decorators = [(0, inspectable_1.Inspect)()];
            _get_rotationAngle_decorators = [(0, inspectable_1.Inspect)()];
            tslib_1.__esDecorate(this, null, _get_type_decorators, { kind: "getter", name: "type", static: false, private: false, access: { has: obj => "type" in obj, get: obj => obj.type }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_topColor_decorators, { kind: "getter", name: "topColor", static: false, private: false, access: { has: obj => "topColor" in obj, get: obj => obj.topColor }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_bottomColor_decorators, { kind: "getter", name: "bottomColor", static: false, private: false, access: { has: obj => "bottomColor" in obj, get: obj => obj.bottomColor }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_rotationAngle_decorators, { kind: "getter", name: "rotationAngle", static: false, private: false, access: { has: obj => "rotationAngle" in obj, get: obj => obj.rotationAngle }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            BackgroundFillGradient = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        /** [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) */
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /**
         * Type of the background fill, always “gradient”
         */
        get type() {
            return this.payload.type;
        }
        /**
         * Top color of the gradient in the RGB24 format
         */
        get topColor() {
            return this.payload.top_color;
        }
        /**
         * Bottom color of the gradient in the RGB24 format
         */
        get bottomColor() {
            return this.payload.bottom_color;
        }
        /**
         * Clockwise rotation angle of the background fill in degrees; 0-359
         */
        get rotationAngle() {
            return this.payload.rotation_angle;
        }
    };
    return BackgroundFillGradient = _classThis;
})();
exports.BackgroundFillGradient = BackgroundFillGradient;
