// @gramio/contexts/dist/structures/chat-boost-removed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatBoostRemoved = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const utils_1 = require("../utils");
const chat_1 = require("./chat");
const index_1 = require("./chat-boost-source/index");
/** This object represents a boost added to a chat or changed. */
let ChatBoostRemoved = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_chat_decorators;
    let _get_id_decorators;
    let _get_removeDate_decorators;
    let _get_source_decorators;
    var ChatBoostRemoved = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_chat_decorators = [(0, inspectable_1.Inspect)()];
            _get_id_decorators = [(0, inspectable_1.Inspect)()];
            _get_removeDate_decorators = [(0, inspectable_1.Inspect)()];
            _get_source_decorators = [(0, inspectable_1.Inspect)()];
            tslib_1.__esDecorate(this, null, _get_chat_decorators, { kind: "getter", name: "chat", static: false, private: false, access: { has: obj => "chat" in obj, get: obj => obj.chat }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_id_decorators, { kind: "getter", name: "id", static: false, private: false, access: { has: obj => "id" in obj, get: obj => obj.id }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_removeDate_decorators, { kind: "getter", name: "removeDate", static: false, private: false, access: { has: obj => "removeDate" in obj, get: obj => obj.removeDate }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_source_decorators, { kind: "getter", name: "source", static: false, private: false, access: { has: obj => "source" in obj, get: obj => obj.source }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            ChatBoostRemoved = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        /** [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) */
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /** Chat which was boosted */
        get chat() {
            return new chat_1.Chat(this.payload.chat);
        }
        /** Unique identifier of the boost */
        get id() {
            return this.payload.boost_id;
        }
        /** Point in time (Unix timestamp) when the boost was removed */
        get removeDate() {
            return this.payload.remove_date;
        }
        /** Source of the removed boost */
        get source() {
            if (this.payload.source.source === "premium") {
                return new index_1.ChatBoostSourcePremium(this.payload.source);
            }
            if (this.payload.source.source === "gift_code") {
                return new index_1.ChatBoostSourceGiftCode(this.payload.source);
            }
            if (this.payload.source.source === "giveaway") {
                return new index_1.ChatBoostSourceGiveaway(this.payload.source);
            }
            throw new TypeError("unknown chat boost source");
        }
    };
    return ChatBoostRemoved = _classThis;
})();
exports.ChatBoostRemoved = ChatBoostRemoved;
(0, utils_1.memoizeGetters)(ChatBoostRemoved, ["chat", "source"]);
