// @gramio/contexts/dist/structures/chat-join-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatJoinRequest = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const utils_1 = require("../utils");
const chat_1 = require("./chat");
const chat_invite_link_1 = require("./chat-invite-link");
const user_1 = require("./user");
/** Represents a join request sent to a chat. */
let ChatJoinRequest = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_chat_decorators;
    let _get_from_decorators;
    let _get_userChatId_decorators;
    let _get_date_decorators;
    let _get_bio_decorators;
    let _get_inviteLink_decorators;
    var ChatJoinRequest = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_chat_decorators = [(0, inspectable_1.Inspect)()];
            _get_from_decorators = [(0, inspectable_1.Inspect)()];
            _get_userChatId_decorators = [(0, inspectable_1.Inspect)()];
            _get_date_decorators = [(0, inspectable_1.Inspect)()];
            _get_bio_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_inviteLink_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            tslib_1.__esDecorate(this, null, _get_chat_decorators, { kind: "getter", name: "chat", static: false, private: false, access: { has: obj => "chat" in obj, get: obj => obj.chat }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_from_decorators, { kind: "getter", name: "from", static: false, private: false, access: { has: obj => "from" in obj, get: obj => obj.from }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_userChatId_decorators, { kind: "getter", name: "userChatId", static: false, private: false, access: { has: obj => "userChatId" in obj, get: obj => obj.userChatId }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_date_decorators, { kind: "getter", name: "date", static: false, private: false, access: { has: obj => "date" in obj, get: obj => obj.date }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_bio_decorators, { kind: "getter", name: "bio", static: false, private: false, access: { has: obj => "bio" in obj, get: obj => obj.bio }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_inviteLink_decorators, { kind: "getter", name: "inviteLink", static: false, private: false, access: { has: obj => "inviteLink" in obj, get: obj => obj.inviteLink }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            ChatJoinRequest = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        /** [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) */
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /** Chat to which the request was sent */
        get chat() {
            return new chat_1.Chat(this.payload.chat);
        }
        /** User that sent the join request */
        get from() {
            return new user_1.User(this.payload.from);
        }
        /** Identifier of a private chat with the user who sent the join request. This number may have more than 32 significant bits and some programming languages may have difficulty/silent defects in interpreting it. But it has at most 52 significant bits, so a 64-bit integer or double-precision float type are safe for storing this identifier. The bot can use this identifier for 24 hours to send messages until the join request is processed, assuming no other administrator contacted the user. */
        get userChatId() {
            return this.payload.user_chat_id;
        }
        /** Date the request was sent in Unix time */
        get date() {
            return this.payload.date;
        }
        /** Bio of the user */
        get bio() {
            return this.payload.bio;
        }
        /** Chat invite link that was used by the user to send the join request */
        get inviteLink() {
            const { invite_link } = this.payload;
            if (!invite_link)
                return undefined;
            return new chat_invite_link_1.ChatInviteLink(invite_link);
        }
    };
    return ChatJoinRequest = _classThis;
})();
exports.ChatJoinRequest = ChatJoinRequest;
(0, utils_1.memoizeGetters)(ChatJoinRequest, ["chat", "from", "inviteLink"]);
