// @gramio/contexts/dist/structures/chat-boost.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatBoost = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const utils_1 = require("../utils");
const index_1 = require("./chat-boost-source/index");
/** This object contains information about a chat boost. */
let ChatBoost = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_id_decorators;
    let _get_addDate_decorators;
    let _get_expirationDate_decorators;
    let _get_source_decorators;
    var ChatBoost = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_id_decorators = [(0, inspectable_1.Inspect)()];
            _get_addDate_decorators = [(0, inspectable_1.Inspect)()];
            _get_expirationDate_decorators = [(0, inspectable_1.Inspect)()];
            _get_source_decorators = [(0, inspectable_1.Inspect)()];
            tslib_1.__esDecorate(this, null, _get_id_decorators, { kind: "getter", name: "id", static: false, private: false, access: { has: obj => "id" in obj, get: obj => obj.id }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_addDate_decorators, { kind: "getter", name: "addDate", static: false, private: false, access: { has: obj => "addDate" in obj, get: obj => obj.addDate }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_expirationDate_decorators, { kind: "getter", name: "expirationDate", static: false, private: false, access: { has: obj => "expirationDate" in obj, get: obj => obj.expirationDate }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_source_decorators, { kind: "getter", name: "source", static: false, private: false, access: { has: obj => "source" in obj, get: obj => obj.source }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            ChatBoost = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        /** [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) */
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /** Unique identifier of the boost */
        get id() {
            return this.payload.boost_id;
        }
        /** Point in time (Unix timestamp) when the chat was boosted */
        get addDate() {
            return this.payload.add_date;
        }
        /** Point in time (Unix timestamp) when the boost will automatically expire, unless the booster's Telegram Premium subscription is prolonged */
        get expirationDate() {
            return this.payload.expiration_date;
        }
        /** Source of the added boost */
        get source() {
            if (this.payload.source.source === "premium") {
                return new index_1.ChatBoostSourcePremium(this.payload.source);
            }
            if (this.payload.source.source === "gift_code") {
                return new index_1.ChatBoostSourceGiftCode(this.payload.source);
            }
            if (this.payload.source.source === "giveaway") {
                return new index_1.ChatBoostSourceGiveaway(this.payload.source);
            }
            throw new TypeError("unknown chat boost source");
        }
    };
    return ChatBoost = _classThis;
})();
exports.ChatBoost = ChatBoost;
(0, utils_1.memoizeGetters)(ChatBoost, ["source"]);
