// @gramio/contexts/dist/structures/chat-shared.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatShared = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const photo_size_1 = require("./photo-size");
/** This object contains information about the chat whose identifier was shared with the bot using a KeyboardButtonRequestChat button. */
let ChatShared = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_requestId_decorators;
    let _get_chatId_decorators;
    let _get_title_decorators;
    let _get_username_decorators;
    let _get_photo_decorators;
    var ChatShared = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_requestId_decorators = [(0, inspectable_1.Inspect)()];
            _get_chatId_decorators = [(0, inspectable_1.Inspect)()];
            _get_title_decorators = [(0, inspectable_1.Inspect)()];
            _get_username_decorators = [(0, inspectable_1.Inspect)()];
            _get_photo_decorators = [(0, inspectable_1.Inspect)()];
            tslib_1.__esDecorate(this, null, _get_requestId_decorators, { kind: "getter", name: "requestId", static: false, private: false, access: { has: obj => "requestId" in obj, get: obj => obj.requestId }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_chatId_decorators, { kind: "getter", name: "chatId", static: false, private: false, access: { has: obj => "chatId" in obj, get: obj => obj.chatId }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_title_decorators, { kind: "getter", name: "title", static: false, private: false, access: { has: obj => "title" in obj, get: obj => obj.title }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_username_decorators, { kind: "getter", name: "username", static: false, private: false, access: { has: obj => "username" in obj, get: obj => obj.username }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_photo_decorators, { kind: "getter", name: "photo", static: false, private: false, access: { has: obj => "photo" in obj, get: obj => obj.photo }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            ChatShared = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        /** [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) */
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /** Identifier of the request */
        get requestId() {
            return this.payload.request_id;
        }
        /** Identifier of the shared chat. This number may have more than 32 significant bits and some programming languages may have difficulty/silent defects in interpreting it. But it has at most 52 significant bits, so a 64-bit integer or double-precision float type are safe for storing this identifier. The bot may not have access to the chat and could be unable to use this identifier, unless the chat is already known to the bot by some other means. */
        get chatId() {
            return this.payload.chat_id;
        }
        /** Title of the chat, if the title was requested by the bot. */
        get title() {
            return this.payload.title;
        }
        /** Username of the chat, if the username was requested by the bot and available. */
        get username() {
            return this.payload.username;
        }
        /** Available sizes of the chat photo, if the photo was requested by the bot. */
        get photo() {
            return this.payload.photo?.map((size) => new photo_size_1.PhotoSize(size));
        }
    };
    return ChatShared = _classThis;
})();
exports.ChatShared = ChatShared;
