// @gramio/contexts/dist/structures/contact.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Contact = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
/** This object represents a phone contact. */
let Contact = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_phoneNumber_decorators;
    let _get_firstName_decorators;
    let _get_lastName_decorators;
    let _get_userId_decorators;
    let _get_vCard_decorators;
    var Contact = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_phoneNumber_decorators = [(0, inspectable_1.Inspect)()];
            _get_firstName_decorators = [(0, inspectable_1.Inspect)()];
            _get_lastName_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_userId_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_vCard_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            tslib_1.__esDecorate(this, null, _get_phoneNumber_decorators, { kind: "getter", name: "phoneNumber", static: false, private: false, access: { has: obj => "phoneNumber" in obj, get: obj => obj.phoneNumber }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_firstName_decorators, { kind: "getter", name: "firstName", static: false, private: false, access: { has: obj => "firstName" in obj, get: obj => obj.firstName }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_lastName_decorators, { kind: "getter", name: "lastName", static: false, private: false, access: { has: obj => "lastName" in obj, get: obj => obj.lastName }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_userId_decorators, { kind: "getter", name: "userId", static: false, private: false, access: { has: obj => "userId" in obj, get: obj => obj.userId }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_vCard_decorators, { kind: "getter", name: "vCard", static: false, private: false, access: { has: obj => "vCard" in obj, get: obj => obj.vCard }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            Contact = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        /** [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) */
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /** Contact's phone number */
        get phoneNumber() {
            return this.payload.phone_number;
        }
        /** Contact's first name */
        get firstName() {
            return this.payload.first_name;
        }
        /** Contact's last name */
        get lastName() {
            return this.payload.last_name;
        }
        /** Contact's user identifier in Telegram */
        get userId() {
            return this.payload.user_id;
        }
        /** Additional data about the contact in the form of a vCard */
        get vCard() {
            return this.payload.vcard;
        }
    };
    return Contact = _classThis;
})();
exports.Contact = Contact;
