// @gramio/contexts/dist/structures/location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Location = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
/** This object represents a point on the map. */
let Location = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_longitude_decorators;
    let _get_latitude_decorators;
    let _get_horizontalAccuracy_decorators;
    let _get_livePeriod_decorators;
    let _get_heading_decorators;
    let _get_proximityAlertRadius_decorators;
    var Location = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_longitude_decorators = [(0, inspectable_1.Inspect)()];
            _get_latitude_decorators = [(0, inspectable_1.Inspect)()];
            _get_horizontalAccuracy_decorators = [(0, inspectable_1.Inspect)()];
            _get_livePeriod_decorators = [(0, inspectable_1.Inspect)()];
            _get_heading_decorators = [(0, inspectable_1.Inspect)()];
            _get_proximityAlertRadius_decorators = [(0, inspectable_1.Inspect)()];
            tslib_1.__esDecorate(this, null, _get_longitude_decorators, { kind: "getter", name: "longitude", static: false, private: false, access: { has: obj => "longitude" in obj, get: obj => obj.longitude }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_latitude_decorators, { kind: "getter", name: "latitude", static: false, private: false, access: { has: obj => "latitude" in obj, get: obj => obj.latitude }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_horizontalAccuracy_decorators, { kind: "getter", name: "horizontalAccuracy", static: false, private: false, access: { has: obj => "horizontalAccuracy" in obj, get: obj => obj.horizontalAccuracy }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_livePeriod_decorators, { kind: "getter", name: "livePeriod", static: false, private: false, access: { has: obj => "livePeriod" in obj, get: obj => obj.livePeriod }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_heading_decorators, { kind: "getter", name: "heading", static: false, private: false, access: { has: obj => "heading" in obj, get: obj => obj.heading }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_proximityAlertRadius_decorators, { kind: "getter", name: "proximityAlertRadius", static: false, private: false, access: { has: obj => "proximityAlertRadius" in obj, get: obj => obj.proximityAlertRadius }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            Location = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        /** [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) */
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /** Longitude as defined by sender */
        get longitude() {
            return this.payload.longitude;
        }
        /** Latitude as defined by sender */
        get latitude() {
            return this.payload.latitude;
        }
        /** The radius of uncertainty for the location, measured in meters; `0-1500` */
        get horizontalAccuracy() {
            return this.payload.horizontal_accuracy;
        }
        /**
         * Time relative to the message sending date,
         * during which the location can be updated, in seconds.
         * For active live locations only.
         */
        get livePeriod() {
            return this.payload.live_period;
        }
        /**
         * The direction in which user is moving, in degrees; `1-360`.
         * For active live locations only.
         */
        get heading() {
            return this.payload.heading;
        }
        /**
         * Maximum distance for proximity alerts about approaching another chat member, in meters.
         * For sent live locations only.
         */
        get proximityAlertRadius() {
            return this.payload.proximity_alert_radius;
        }
    };
    return Location = _classThis;
})();
exports.Location = Location;
