// @gramio/contexts/dist/structures/order-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderInfo = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const utils_1 = require("../utils");
const shipping_address_1 = require("./shipping-address");
/** This object represents information about an order. */
let OrderInfo = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_name_decorators;
    let _get_phoneNumber_decorators;
    let _get_email_decorators;
    let _get_shippingAddress_decorators;
    var OrderInfo = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_name_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_phoneNumber_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_email_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_shippingAddress_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            tslib_1.__esDecorate(this, null, _get_name_decorators, { kind: "getter", name: "name", static: false, private: false, access: { has: obj => "name" in obj, get: obj => obj.name }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_phoneNumber_decorators, { kind: "getter", name: "phoneNumber", static: false, private: false, access: { has: obj => "phoneNumber" in obj, get: obj => obj.phoneNumber }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_email_decorators, { kind: "getter", name: "email", static: false, private: false, access: { has: obj => "email" in obj, get: obj => obj.email }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_shippingAddress_decorators, { kind: "getter", name: "shippingAddress", static: false, private: false, access: { has: obj => "shippingAddress" in obj, get: obj => obj.shippingAddress }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            OrderInfo = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        /** [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) */
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /** User name */
        get name() {
            return this.payload.name;
        }
        /** User's phone number */
        get phoneNumber() {
            return this.payload.phone_number;
        }
        /** User email */
        get email() {
            return this.payload.email;
        }
        /** User shipping address */
        get shippingAddress() {
            const { shipping_address } = this.payload;
            if (!shipping_address)
                return undefined;
            return new shipping_address_1.ShippingAddress(shipping_address);
        }
    };
    return OrderInfo = _classThis;
})();
exports.OrderInfo = OrderInfo;
(0, utils_1.memoizeGetters)(OrderInfo, ["shippingAddress"]);
