// @gramio/contexts/dist/structures/paid-media-preview.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaidMediaPreview = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
/**
 * The paid media isn't available before the payment.
 *
 * [Documentation](https://core.telegram.org/bots/api/#paidmediapreview)
 */
let PaidMediaPreview = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_type_decorators;
    let _get_width_decorators;
    let _get_height_decorators;
    let _get_duration_decorators;
    var PaidMediaPreview = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_type_decorators = [(0, inspectable_1.Inspect)()];
            _get_width_decorators = [(0, inspectable_1.Inspect)()];
            _get_height_decorators = [(0, inspectable_1.Inspect)()];
            _get_duration_decorators = [(0, inspectable_1.Inspect)()];
            tslib_1.__esDecorate(this, null, _get_type_decorators, { kind: "getter", name: "type", static: false, private: false, access: { has: obj => "type" in obj, get: obj => obj.type }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_width_decorators, { kind: "getter", name: "width", static: false, private: false, access: { has: obj => "width" in obj, get: obj => obj.width }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_height_decorators, { kind: "getter", name: "height", static: false, private: false, access: { has: obj => "height" in obj, get: obj => obj.height }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_duration_decorators, { kind: "getter", name: "duration", static: false, private: false, access: { has: obj => "duration" in obj, get: obj => obj.duration }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            PaidMediaPreview = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /**
         * Type of the paid media, always “preview”
         */
        get type() {
            return this.payload.type;
        }
        /**
         * *Optional*. Media width as defined by the sender
         */
        get width() {
            return this.payload.width;
        }
        /**
         * *Optional*. Media height as defined by the sender
         */
        get height() {
            return this.payload.height;
        }
        /**
         * *Optional*. Duration of the media in seconds as defined by the sender
         */
        get duration() {
            return this.payload.duration;
        }
    };
    return PaidMediaPreview = _classThis;
})();
exports.PaidMediaPreview = PaidMediaPreview;
