// @gramio/contexts/dist/structures/pre-checkout-query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreCheckoutQuery = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const utils_1 = require("../utils");
const order_info_1 = require("./order-info");
const user_1 = require("./user");
/** This object contains information about an incoming pre-checkout query. */
let PreCheckoutQuery = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_id_decorators;
    let _get_from_decorators;
    let _get_senderId_decorators;
    let _get_currency_decorators;
    let _get_totalAmount_decorators;
    let _get_invoicePayload_decorators;
    let _get_shippingOptionId_decorators;
    let _get_orderInfo_decorators;
    var PreCheckoutQuery = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_id_decorators = [(0, inspectable_1.Inspect)()];
            _get_from_decorators = [(0, inspectable_1.Inspect)()];
            _get_senderId_decorators = [(0, inspectable_1.Inspect)()];
            _get_currency_decorators = [(0, inspectable_1.Inspect)()];
            _get_totalAmount_decorators = [(0, inspectable_1.Inspect)()];
            _get_invoicePayload_decorators = [(0, inspectable_1.Inspect)()];
            _get_shippingOptionId_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_orderInfo_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            tslib_1.__esDecorate(this, null, _get_id_decorators, { kind: "getter", name: "id", static: false, private: false, access: { has: obj => "id" in obj, get: obj => obj.id }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_from_decorators, { kind: "getter", name: "from", static: false, private: false, access: { has: obj => "from" in obj, get: obj => obj.from }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_senderId_decorators, { kind: "getter", name: "senderId", static: false, private: false, access: { has: obj => "senderId" in obj, get: obj => obj.senderId }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_currency_decorators, { kind: "getter", name: "currency", static: false, private: false, access: { has: obj => "currency" in obj, get: obj => obj.currency }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_totalAmount_decorators, { kind: "getter", name: "totalAmount", static: false, private: false, access: { has: obj => "totalAmount" in obj, get: obj => obj.totalAmount }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_invoicePayload_decorators, { kind: "getter", name: "invoicePayload", static: false, private: false, access: { has: obj => "invoicePayload" in obj, get: obj => obj.invoicePayload }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_shippingOptionId_decorators, { kind: "getter", name: "shippingOptionId", static: false, private: false, access: { has: obj => "shippingOptionId" in obj, get: obj => obj.shippingOptionId }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_orderInfo_decorators, { kind: "getter", name: "orderInfo", static: false, private: false, access: { has: obj => "orderInfo" in obj, get: obj => obj.orderInfo }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            PreCheckoutQuery = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        /** [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) */
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /** Unique query identifier */
        get id() {
            return this.payload.id;
        }
        /** User who sent the query */
        get from() {
            return new user_1.User(this.payload.from);
        }
        /** Sender ID */
        get senderId() {
            return this.from.id;
        }
        /** Three-letter ISO 4217 currency code */
        get currency() {
            return this.payload.currency;
        }
        /**
         * Total price in the smallest units of the currency
         * (integer, not float/double). For example, for a price of
         * `US$ 1.45` pass `amount = 145`. See the `exp` parameter in
         * [currencies.json](https://core.telegram.org/bots/payments/currencies.json),
         * it shows the number of digits past the decimal point for each currency
         * (2 for the majority of currencies).
         */
        get totalAmount() {
            return this.payload.total_amount;
        }
        /** Bot specified invoice payload */
        get invoicePayload() {
            return this.payload.invoice_payload;
        }
        /** Identifier of the shipping option chosen by the user */
        get shippingOptionId() {
            return this.payload.shipping_option_id;
        }
        /** Order info provided by the user */
        get orderInfo() {
            const { order_info } = this.payload;
            if (!order_info)
                return undefined;
            return new order_info_1.OrderInfo(order_info);
        }
    };
    return PreCheckoutQuery = _classThis;
})();
exports.PreCheckoutQuery = PreCheckoutQuery;
(0, utils_1.memoizeGetters)(PreCheckoutQuery, ["from", "orderInfo"]);
