// @gramio/contexts/dist/structures/shipping-address.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShippingAddress = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
/** This object represents a shipping address. */
let ShippingAddress = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_countryCode_decorators;
    let _get_state_decorators;
    let _get_city_decorators;
    let _get_firstStreetLine_decorators;
    let _get_secondStreetLine_decorators;
    let _get_postCode_decorators;
    var ShippingAddress = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_countryCode_decorators = [(0, inspectable_1.Inspect)()];
            _get_state_decorators = [(0, inspectable_1.Inspect)()];
            _get_city_decorators = [(0, inspectable_1.Inspect)()];
            _get_firstStreetLine_decorators = [(0, inspectable_1.Inspect)()];
            _get_secondStreetLine_decorators = [(0, inspectable_1.Inspect)()];
            _get_postCode_decorators = [(0, inspectable_1.Inspect)()];
            tslib_1.__esDecorate(this, null, _get_countryCode_decorators, { kind: "getter", name: "countryCode", static: false, private: false, access: { has: obj => "countryCode" in obj, get: obj => obj.countryCode }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_state_decorators, { kind: "getter", name: "state", static: false, private: false, access: { has: obj => "state" in obj, get: obj => obj.state }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_city_decorators, { kind: "getter", name: "city", static: false, private: false, access: { has: obj => "city" in obj, get: obj => obj.city }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_firstStreetLine_decorators, { kind: "getter", name: "firstStreetLine", static: false, private: false, access: { has: obj => "firstStreetLine" in obj, get: obj => obj.firstStreetLine }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_secondStreetLine_decorators, { kind: "getter", name: "secondStreetLine", static: false, private: false, access: { has: obj => "secondStreetLine" in obj, get: obj => obj.secondStreetLine }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_postCode_decorators, { kind: "getter", name: "postCode", static: false, private: false, access: { has: obj => "postCode" in obj, get: obj => obj.postCode }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            ShippingAddress = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        /** [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) */
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /** ISO 3166-1 alpha-2 country code */
        get countryCode() {
            return this.payload.country_code;
        }
        /** State, if applicable */
        get state() {
            return this.payload.state;
        }
        /** City */
        get city() {
            return this.payload.city;
        }
        /** First line for the address */
        get firstStreetLine() {
            return this.payload.street_line1;
        }
        /** Second line for the address */
        get secondStreetLine() {
            return this.payload.street_line2;
        }
        /** Address post code */
        get postCode() {
            return this.payload.post_code;
        }
    };
    return ShippingAddress = _classThis;
})();
exports.ShippingAddress = ShippingAddress;
