// @gramio/contexts/dist/structures/venue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Venue = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const utils_1 = require("../utils");
const location_1 = require("./location");
/** This object represents a venue. */
let Venue = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _instanceExtraInitializers = [];
    let _get_location_decorators;
    let _get_title_decorators;
    let _get_address_decorators;
    let _get_foursquareId_decorators;
    let _get_foursquareType_decorators;
    let _get_googlePlaceId_decorators;
    let _get_googlePlaceType_decorators;
    var Venue = class {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _get_location_decorators = [(0, inspectable_1.Inspect)()];
            _get_title_decorators = [(0, inspectable_1.Inspect)()];
            _get_address_decorators = [(0, inspectable_1.Inspect)()];
            _get_foursquareId_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_foursquareType_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_googlePlaceId_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _get_googlePlaceType_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            tslib_1.__esDecorate(this, null, _get_location_decorators, { kind: "getter", name: "location", static: false, private: false, access: { has: obj => "location" in obj, get: obj => obj.location }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_title_decorators, { kind: "getter", name: "title", static: false, private: false, access: { has: obj => "title" in obj, get: obj => obj.title }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_address_decorators, { kind: "getter", name: "address", static: false, private: false, access: { has: obj => "address" in obj, get: obj => obj.address }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_foursquareId_decorators, { kind: "getter", name: "foursquareId", static: false, private: false, access: { has: obj => "foursquareId" in obj, get: obj => obj.foursquareId }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_foursquareType_decorators, { kind: "getter", name: "foursquareType", static: false, private: false, access: { has: obj => "foursquareType" in obj, get: obj => obj.foursquareType }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_googlePlaceId_decorators, { kind: "getter", name: "googlePlaceId", static: false, private: false, access: { has: obj => "googlePlaceId" in obj, get: obj => obj.googlePlaceId }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_googlePlaceType_decorators, { kind: "getter", name: "googlePlaceType", static: false, private: false, access: { has: obj => "googlePlaceType" in obj, get: obj => obj.googlePlaceType }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            Venue = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            this.payload = payload;
        }
        /** [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) */
        get [Symbol.toStringTag]() {
            return this.constructor.name;
        }
        /** Venue location */
        get location() {
            return new location_1.Location(this.payload.location);
        }
        /** Name of the venue */
        get title() {
            return this.payload.title;
        }
        /** Address of the venue */
        get address() {
            return this.payload.address;
        }
        /** Foursquare identifier of the venue */
        get foursquareId() {
            return this.payload.foursquare_id;
        }
        /** Foursquare type of the venue */
        get foursquareType() {
            return this.payload.foursquare_type;
        }
        /** Google Places identifier of the venue */
        get googlePlaceId() {
            return this.payload.google_place_id;
        }
        /**
         * Google Places type of the venue.
         * (See [supported types](https://developers.google.com/places/web-service/supported_types).)
         */
        get googlePlaceType() {
            return this.payload.google_place_type;
        }
    };
    return Venue = _classThis;
})();
exports.Venue = Venue;
(0, utils_1.memoizeGetters)(Venue, ["location"]);
