// @gramio/contexts/dist/structures/chat-boost-source/giveaway.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatBoostSourceGiveaway = void 0;
const tslib_1 = require("tslib");
const inspectable_1 = require("inspectable");
const user_1 = require("../user");
const utils_1 = require("../../utils");
const chat_boost_source_1 = require("./chat-boost-source");
/** The boost was obtained by the creation of a Telegram Premium giveaway. This boosts the chat 4 times for the duration of the corresponding Telegram Premium subscription. */
let ChatBoostSourceGiveaway = (() => {
    let _classDecorators = [(0, inspectable_1.Inspectable)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = chat_boost_source_1.ChatBoostSource;
    let _instanceExtraInitializers = [];
    let _get_source_decorators;
    let _get_user_decorators;
    let _isUnclaimed_decorators;
    var ChatBoostSourceGiveaway = class extends _classSuper {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _get_source_decorators = [(0, inspectable_1.Inspect)()];
            _get_user_decorators = [(0, inspectable_1.Inspect)({ nullable: false })];
            _isUnclaimed_decorators = [(0, inspectable_1.Inspect)({ compute: true, nullable: false })];
            tslib_1.__esDecorate(this, null, _get_source_decorators, { kind: "getter", name: "source", static: false, private: false, access: { has: obj => "source" in obj, get: obj => obj.source }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _get_user_decorators, { kind: "getter", name: "user", static: false, private: false, access: { has: obj => "user" in obj, get: obj => obj.user }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(this, null, _isUnclaimed_decorators, { kind: "method", name: "isUnclaimed", static: false, private: false, access: { has: obj => "isUnclaimed" in obj, get: obj => obj.isUnclaimed }, metadata: _metadata }, null, _instanceExtraInitializers);
            tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            ChatBoostSourceGiveaway = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            tslib_1.__runInitializers(_classThis, _classExtraInitializers);
        }
        payload = tslib_1.__runInitializers(this, _instanceExtraInitializers);
        constructor(payload) {
            super(payload);
            this.payload = payload;
        }
        /** Source of the boost, always `giveaway` */
        get source() {
            return this.payload.source;
        }
        get giveawayMessageId() {
            return this.payload.giveaway_message_id;
        }
        /** User that boosted the chat */
        get user() {
            const { user } = this.payload;
            if (!user)
                return undefined;
            return new user_1.User(user);
        }
        /** `true`, if the giveaway was completed, but there was no user to win the prize */
        isUnclaimed() {
            return this.payload.is_unclaimed;
        }
    };
    return ChatBoostSourceGiveaway = _classThis;
})();
exports.ChatBoostSourceGiveaway = ChatBoostSourceGiveaway;
(0, utils_1.memoizeGetters)(ChatBoostSourceGiveaway, ["user"]);
